### ISSUE TITLE
testing results poor after model converging

### ISSUE THREAD

**Issue by antoniosimunovic** [emedvedev/attention-ocr#4]
Hi,

I've trained the model on a Synth 90k dataset and after ~442800 iterations reported perplexity is ~1.05, precision ~90% and loss in the 0.01-0.1 range. But when I run test command on same dataset numbers are much worse. Here is statistics after 1000 test steps: accuracy ~35%, perplexity ~170528.

This is train command:

`aocr train datasets/synth90k_train.tfrecord --max-width 200 --max-height 31 --max-prediction 30`

Ant this is test command:

`aocr test datasets/synth90k_train.tfrecord --max-width 200 --max-height 31 --max-prediction 30`

Any idea why such poor results?

**Comment by antoniosimunovic** [comment]
Ok, just realized that there is a difference in calculating number of correct recognitions during training and testing: in training phase number of incorrect character recognitions is calculated using the levenshtein distance of prediction and ground truth label, but during test phase exact match is required for example to be correct (zero levenshtein distance).

Still, reported loss is much higher during testing: in the 10-15 range. As a consequence of that, model perplexitiy is also high: ~500000.

**Comment by emedvedev** [comment]
Hi,

There's a couple things to note here:

1. Loss and perplexity counts during testing are broken. I haven't looked too deep into it, but for now you might consider accuracy as your primary metric, not perplexity/loss.

2. You're right about Levenshtein during training vs full match during testing: it was important for my use-case where even the distance of 1 would invalidate the result. I'll introduce a CLI flag to calculate Levenshtein and not full match during testing as well.

As for low accuracy, I think you may be overfitting. Try inspecting the training/testing process through TensorBoard or play with hyperparameters. I'll also try Synth90k and see if I can reproduce the low score, but that may not happen for a week or so—this project is more of a hobby than a job.

**Comment by emedvedev** [comment]
Google ML Cloud fails on me today for some reason, so I've tested the results on a half-trained model (2100 steps with a batch size of 512). The training accuracy is roughly equal to the testing accuracy: ~79% correctness on training, ~76% on testing).

I'll publish full results once the model reaches convergence, but so far there's a few observations about training it on Synth90k: 

1. I've had to limit the training and testing sets to the images of up to 302 pixels in width because it gets hella slow (so I'm actually using around 90% of the dataset).

2. The original annotations are lexicon-based (made to recognize words as a whole), and this model is attention-based (made to recognize individual characters). I imagine the word-based models would be faster and more accurate, but their application is obviously limited to the words they are trained to recognize. I've had to modify the annotations to train this model on Synth90k.

**Comment by emedvedev** [comment]
@antoniosimunovic I've released a few bug fixes in the latest version, and after moderate amount of training (~8000 iterations with a batch size of 512) I get ~93% accuracy during testing. More training and better tuning of hyperparameters could definitely increase that result.

Closing the issue, as the originally reported bug is now fixed. Thanks a lot for giving the model a try!

**Comment by emedvedev** [comment]
Trained checkpoint: https://www.dropbox.com/s/y9duuc4bzxwuh29/synth90k.zip?dl=0
